### DISCUSSION TITLE
Delete blocks from DXF file

### DISCUSSION THREAD

**Discussion by preethamam** [mozman/ezdxf#1184] [Q&A]
Hello ezdxf community:

I am new to `ezdxf`. I am trying to delete all the blocks from DXF file and save the same. However, I get an error:
`AttributeError: 'BlockRecord' object has no attribute 'dxf'`. Here is the link to the [DXF file](https://1drv.ms/u/s!AlFYM4jwmzqrvaV7U2QTzjkyfijzaA?e=tMEGt2).

Below is the code:
```python
doc = ezdxf.readfile(file_path)
msp = doc.modelspace()
blocks = doc.blocks
block_names = [block.dxf.name for block in blocks]
for block_name in block_names:
  for entity in msp.query('INSERT'):
    if entity.dxf.name == block_name:
        msp.delete_entity(entity)
        print(f"Deleted INSERT entity referencing block: {block_name}")

for block_name in block_names:
  try:
      doc.blocks.delete_block(name=block_name, safe=False)
      print(f"Deleted block definition: {block_name}")
  except ezdxf.DXFBlockInUseError:
      print(f"Block {block_name} is still in use and could not be deleted.")

doc.saveas("blocks_deleted.dxf")
```

Any help is appreciated.

**Comment by mozman** [comment]
If you are new to DXF, do not blindly delete objects in the DXF file! The DXF file format is complex and internal structures can easily be destroyed.

I will look into the problem when I have time.

**Comment by mozman** [comment]
So, I've looked into your "problem" and I can't really help you because you're trying to shoot yourself in the foot and `ezdxf` allows that.

You have bypassed one of the few safeguards in the library:
```python

doc.blocks.delete_block(name=block_name, safe=False)
#                                        ^^^^^^^^^^
```

The current state of the method allows you to destroy required layouts even when `safe` is `True`. This bug is now fixed, but don't worry your code still destroys the drawing by deleting custom dimension arrow heads which are not protected by the safeguards.

This version deletes only blocks that are referenced in modelspace:

```python
import ezdxf

doc = ezdxf.readfile("0_Tower_B_Coord_25_Floor_structural.dxf")
msp = doc.modelspace()

referenced_blocks = set()
for insert in msp.query("INSERT"):
    referenced_blocks.add(insert.dxf.name)
    msp.delete_entity(insert)
    

for block_name in referenced_blocks:
    try:
        doc.blocks.delete_block(name=block_name, safe=True)
    except ezdxf.DXFBlockInUseError:
        pass

# Safeguard for users who do not want to export destroyed DXF documents
report = doc.audit()
if not report.has_errors:
    doc.saveas("blocks_deleted.dxf")
else:
    report.print_error_report()
```

**Comment by mozman** [comment]
This is a different solution that utilizes the [`ezdxf.blkrefs`](https://ezdxf.mozman.at/docs/blkrefs.html) module to detect unused blocks. The class `BlockReferenceCounter` counts all block references in a DXF document, taking into account all references by any entity or resource (DIMSYTLE, MLEADERSTYLE).

```Python
import ezdxf
from ezdxf import blkrefs

doc = ezdxf.readfile("0_Tower_B_Coord_25_Floor_structural.dxf")
ref_counter = blkrefs.BlockReferenceCounter(doc)

unreferenced_blocks = set()
for block in doc.blocks:
    if not block.is_alive or block.is_any_layout:
        continue
    count = ref_counter.by_name(block.name)
    if count == 0:
        # DO NOT DELETE BLOCK WHILE ITERATING!
        unreferenced_blocks.add(block.name)

for block_name in unreferenced_blocks:
    # The blkrefs.BlockReferenceCounter counts ALL KNOWN references, it should be safe 
    # to destroy these blocks without safeguards (no guarantee for that!):
    doc.blocks.delete_block(block_name, safe=False)

report = doc.audit()
if not report.has_errors:
    doc.saveas("unused_blocks_purged.dxf")
else:
    report.print_error_report()
```

This code reduces the filesize from 86.5MB to 2.3 MB, like the nested PURGE command in BricsCAD.

**Reply by cococosusu** [reply]
Wow, that's a great feature! Thank you!

**Reply by preethamam** [reply]
@mozman, thank you for the solutions! 

I used both of your solutions. Solution 1:  "blocks that are referenced in modelspace" delete the blocks `Window_1`, `Window_2`, `Window_3` ..., as they are the only ones referenced. In solution 2: "Delete unused blocks" should delete everything except the `Window_1`, `Window_2`, `Window_3` ... However, I see that there are other blocks present even after deleting. 

Please note that I have installed the `ezdxf` version `1.3.4`.

I am I missing anything? Any help is appreciated.
